cvmfs/export_plugin/fs_traversal.cc

```diff
@@ -0,0 +1,793 @@
+
+    if (next_dir->IsTerminateJob())
+      break;
+
+    if (!Sync(next_dir->dir, src, dest, next_dir->recursive, pstats, do_fsck)) {
+      LogCvmfs(kLogCvmfs, kLogStderr,
+        "File %s failed to copy\n", next_dir->dir);
+      return false;
+    }
+
+    delete next_dir;
+  }
+  return true;
+}
+
+struct MainWorkerContext {
+  struct fs_traversal *src_fs;
+  struct fs_traversal *dest_fs;
+  perf::Statistics *pstats;
+  int parallel;
+};
+
+struct MainWorkerSpecificContext {
+  struct MainWorkerContext *mwc;
+  int num_thread;
+};
+
+static void *MainWorker(void *data) {
+  MainWorkerSpecificContext *sc = static_cast<MainWorkerSpecificContext*>(data);
+  MainWorkerContext *mwc = sc->mwc;
+  perf::Counter *files_transferred
+    = mwc->pstats->Lookup(SHRINKWRAP_STAT_FILE_COUNT);
+  time_t last_print_time = 0;
+  if (sc->num_thread == 0) {
+    last_print_time = time(NULL);
+  }
+
+  while (1) {
+    if (sc->num_thread == 0 && time(NULL)-last_print_time > 10) {
+      LogCvmfs(kLogCvmfs, kLogStdout,
+        "%s",
+        mwc->pstats->PrintList(perf::Statistics::kPrintSimple).c_str());
+      last_print_time = time(NULL);
+    }
+    FileCopy next_copy;
+    pthread_mutex_lock(&lock_pipe);
+    ReadPipe(pipe_chunks[0], &next_copy, sizeof(next_copy));
+    pthread_mutex_unlock(&lock_pipe);
+    if (next_copy.IsTerminateJob())
+      break;
+
+    if (!next_copy.src || !next_copy.dest) {
+      continue;
+    }
+    if (!copyFile(mwc->src_fs, next_copy.src, mwc->dest_fs,
+                  next_copy.dest, mwc->pstats)) {
+      LogCvmfs(kLogCvmfs, kLogStderr,
+      "File %s failed to copy\n", next_copy.src);
+    }
+    files_transferred->Inc();
+
+    free(next_copy.src);
+    free(next_copy.dest);
+
+    atomic_dec64(&copy_queue);
+  }
+  return NULL;
+}
+
+perf::Statistics *GetSyncStatTemplate() {
+  perf::Statistics *result = new perf::Statistics();
+  result->Register(SHRINKWRAP_STAT_BYTE_COUNT,
+    "The number of bytes transfered from the source to the destination");
+  result->Register(SHRINKWRAP_STAT_FILE_COUNT,
+    "The number of files transfered from the source to the destination");
+  result->Register(SHRINKWRAP_STAT_SRC_ENTRIES,
+    "The number of file system entries processed in the source");
+  result->Register(SHRINKWRAP_STAT_DEST_ENTRIES,
+    "The number of file system entries processed in the destination");
+  result->Register(SHRINKWRAP_STAT_DEDUPED_FILES,
+    "The number of files not copied thanks to deduplication");
+  result->Register(SHRINKWRAP_STAT_DEDUPED_BYTES,
+    "The number of bytes not copied thanks to deduplication");
+  return result;
+}
+
+int SyncInit(
+  struct fs_traversal *src,
+  struct fs_traversal *dest,
+  const char *base,
+  const char *spec,
+  unsigned parallel,
+  unsigned retries
+) {
+  num_parallel_ = parallel;
+  retries_ = retries;
+
+  perf::Statistics *pstats = GetSyncStatTemplate();
+
+  // Initialization
+  atomic_init64(&overall_copies);
+  atomic_init64(&overall_new);
```

Comment:
`overall_copies` and `overall_new` are unused

Reply:
I will remove these. Similar variables were used in Swissknife Pull, but we can track the same things using the statistics structure.